### ISSUE TITLE
IDE0078 - Use pattern matching removes necessary side-effects in some cases

### ISSUE THREAD

**Issue by bartdesmet** [dotnet/roslyn#50462]
Consider the following fragment from some parser using a `reader` to get characters:

```csharp
case 'n':
    // null
    {
        reader.Read();
        if (reader.Read() != 'u' || reader.Read() != 'l' || reader.Read() != 'l')
            throw new ParseException(/* omitted details */);
    }
    break;
```

IDE0078 suggests to rewrite this as:

```csharp
case 'n':
    // null
    {
        reader.Read();
        if (reader.Read() is not 'u' or not 'l' or not 'l')
            throw new ParseException(/* omitted details */);
    }
    break;
```

which removes the `Read()` side-effect, and is immediately followed by another diagnostic that indicates that the pattern always matches.

Some more constraints may have to be enforced on the common expression (here `reader.Read()`) to filter out false positives where this rewrite is unsafe. Ideally, the expression is "pure" but that may be too stringent (e.g. property getters would have to be considered as potentially impure).

**Comment by CyrusNajmabadi** [comment]
@alrz ?

**Comment by alrz** [comment]
The fix I think is to gate `not X or not Y` combination as it always evaluates to true.

As for side-effects, we decided to not filter things out since that would exclude lots of useful cases as well.

**Comment by CyrusNajmabadi** [comment]
I think we may need to warn in the preview though here to let people know semantics may change.

**Comment by Blokyk** [comment]
Any update on this ? I was doing some refactoring and just accepted that code fix without thinking too much about it (since most of the other occurrences of this were valid), refactored some more, and then spent about 15 minutes debugging to find that this tiny change was the culprit.

As for the suggested "side-effect warning," I personally think it would create a bit too much clutter and/or feel out of place. Maybe it's possible to create two different codefixes, one for the side-effect-free version and one for the unsafe version, maybe even having the first in IDE and the second in CA ?

**Comment by CyrusNajmabadi** [comment]
> Any update on this 

No there has been no update on this.  

> Maybe it's possible to create two different codefixes, one for the side-effect-free version and one for the unsafe version,

**Comment by Blokyk** [comment]
Should I create a separate issue for it ? I'm not sure I'd be up for the task of adding a new code fix in roslyn directly tbh 😅

**Comment by CyrusNajmabadi** [comment]
Thanks @Blokyk !

**Comment by Blokyk** [comment]
Thank you most of all for all the help :) Did you see my last message in the PR thread btw ? Something about being able to disable the warning for the unsafe version but leave the safe one alone

**Comment by CyrusNajmabadi** [comment]
> Something about being able to disable the warning for the unsafe version but leave the safe one alone

I think it's better if we just show them, and allow the user to disable in teh rare cases where these expressions are sideeffecting. Thanks!
